### ISSUE TITLE
Allow custom native IntegrationTestRunners for custom test method channels

### ISSUE THREAD

**Issue by JulianBissekkou** [flutter/flutter#113519]
## Use case

I would like to interact with the native test sdk for be able to test native features like Face ID, notification dialogs or interactions like moving the app into background or foreground.
This can be implemented by a platform channel that is created on the native side inside our `MainActivityTest` on Android or an integration test on iOS.

The Flutter integration test sdk already does this in `FlutterTestRunner` and `FLTIntegrationTestRunner`.

## Proposal

Add a way to register an own test runner that can setup method channels for native testing.
On Android we can simply extend the default `FlutterTestRunner` but on iOS I don't see a way of extending `FLTIntegrationTestRunner` and adding my own test plugin.

**Comment by jmagman** [comment]
This sounds like a duplicate of https://github.com/flutter/flutter/issues/101278 ?  There's nothing stopping you from writing black-box XCUITests against your Flutter app.

From https://github.com/flutter/flutter/issues/101278#issuecomment-1087945640

> `integration_test` and `XCTest` are for unit/integration testing, you can mock out different layers of your app and interact with the classes and state directly. `XCUITest` is black-box testing that only allows you to interact with an app as a user would via accessibility (pressing buttons, typing text, confirming something is on the screen). You can easily create `XCUITest` black box tests for your Flutter app the same way you can with any iOS or macOS app.
> 
> Examples: https://github.com/flutter/flutter/blob/7428ab66b4e79b695ffc03a6e15a2d18b019aa7f/dev/integration_tests/ios_host_app/FlutterUITests/FlutterUITests.m https://github.com/flutter/plugins/blob/main/packages/image_picker/image_picker/example/ios/RunnerUITests/ImagePickerFromLimitedGalleryUITests.m https://github.com/flutter/plugins/blob/4ebfcb53b93144cbe4a7e848c992fb394a697641/packages/video_player/video_player_avfoundation/example/ios/RunnerUITests/VideoPlayerUITests.m https://github.com/flutter/plugins/blob/411f09e3d8b13cb51eb14d2e05bab8583622dcaa/packages/webview_flutter/webview_flutter_wkwebview/example/ios/RunnerUITests/FLTWebViewUITests.m
> 
> We could add some documentation for this somewhere, but not in `integration_test` since that would never be appropriate for black-box testing.

**Comment by JulianBissekkou** [comment]
Are we able to execute an integration test inside my XCUITest? I don't see an option to do this inside my XCUITest.

Here is a specific scenario to illustrate what I want to achive:
Run my flutter integration test that clicks some flutter widgets and do some assertions. At some point in my test I would like to put my app into the background using a method channel that I call from my integration test in dart. The method channel needs to be registered in my XCUITest.

**Comment by jmagman** [comment]
"The method channel needs to be registered in my XCUITest" this can't be done directly with XCUITests, and it has nothing to do with Flutter.  With a vanilla Xcode app and XCUITest you can't call methods in your app, you can only interact with the UI and do limited lifecycle things with `XCUIApplication`.

> `XCUITest` is black-box testing that only allows you to interact with an app as a user would via accessibility (pressing buttons, typing text, confirming something is on the screen)

One way to do this instead is to have a button that calls the method channel that puts the app into the background.  Then your XCUITest presses the button.

**Comment by JulianBissekkou** [comment]
Okay then this is not a duplicate of https://github.com/flutter/flutter/issues/101278 and this requires adjustments in `IntegrationTestIosTest` where `INTEGRATION_TEST_IOS_RUNNER` is defined.

**Comment by jmagman** [comment]
> This can be implemented by a platform channel that is created on the native side inside our MainActivityTest on Android or an integration test on iOS.

This is true, I'm still confused what's being asked for.  Can you be really specific about what tests you want to write that are possible on Android but not iOS?

> Add a way to register an own test runner that can setup method channels for native testing.

What _exactly_ is keeping you from doing this on the native side?

**Comment by bartekpacia** [comment]
I think that @JulianBissekkou means that when using XCUITest, the tests and the app run in separate processes, which means there is no way to facilitate communications between them using a method channel (because we're crossing the process boundary).

The `integration_test` package uses method channels to send test results from Dart side to native side – this works fine on Android (because on Android, the instrumentation and the app both run in the same process), and works fine on iOS as well, but only when using XCTest ("Unit Testing Bundle" target in Xcode). If we wanted to use XCUITest ("UI Testing Bundle" target in Xcode), we can't do that, because (as far as I know) it's not possible to set up a method channel between two separate processes.

Instead of using method channels, we could use some form of RPC, e.g [eDistantObject](https://github.com/google/eDistantObject) (which is what [EarlGrey 2](https://github.com/google/EarlGrey/tree/earlgrey2) uses).

**Comment by jmagman** [comment]
> I think that @JulianBissekkou means that when using XCUITest, the tests and the app run in separate processes, which means there is no way to facilitate communications between them using a method channel (because we're crossing the process boundary).

That's exactly what I was saying as well:

> "The method channel needs to be registered in my XCUITest" this can't be done directly with XCUITests, and it has nothing to do with Flutter. With a vanilla Xcode app and XCUITest you can't call methods in your app, you can only interact with the UI and do limited lifecycle things with `XCUIApplication`.

If Flutter wasn't involved, let's say you had a vanilla iOS or macOS framework embedded in your app that did something special during app backgrounding, and you wanted to test with XCUITest.  There's no way to directly interact with the framework from an XCUITest, the app is a complete black-box and you can only interact with the UI.  You can background it and check the behavior, but you can't actually call or mock methods in the framework.

Now bring Flutter and methods channels into play--it's the same thing, you can't directly interact with the message channels because you don't have access to poke at the actual app process the way you do with a XCTest.

So I'm saying this is Apple's by-design limitation of XCUITest that has nothing to do with Flutter.  I've suggested some ways to work around this:

> One way to do this instead is to have a button that calls the method channel that puts the app into the background. Then your XCUITest presses the button.

You could also use EarlGrey, as mentioned.  Though here's an example of an EarlyGrey test we had that I re-wrote as XCUITest: https://github.com/flutter/flutter/pull/61007

**Comment by bartekpacia** [comment]
> Now bring Flutter and methods channels into play--it's the same thing, you can't directly interact with the message channels because you don't have access to poke at the actual app process the way you do with a XCTest.

Yup, so some form of RPC is needed. The EarlGrey library you mentioned uses eDistantObject for this, but any form of RPC 
should work for this case, e.g gRPC.

FWIW, we're working on this (this=`integration_test`  with native test reporting + XCUITest for automation superpowers) in [our Patrol test framework](https://github.com/leancodepl/patrol) - see [this issue](https://github.com/leancodepl/patrol/issues/661).

**Comment by jmagman** [comment]
So is the feature request to implement something in `integration_test` that wraps EarlGrey/eDistantObject/gRPC to allow you to poke the message channels from XCUITest (black-box use case as described in https://github.com/google/EarlGrey/tree/earlgrey2#earlgrey-20)?  

Here's an example of a Flutter test using EarlGrey by the way https://github.com/flutter/flutter/blob/08a2635e2b725c951ff0471eb9c556b375aa5d7a/dev/integration_tests/ios_add2app_life_cycle/ios_add2appTests/IntegrationTests.m
